hydroflow_lang/src/graph/flat_graph_builder.rs

```diff
@@ -888,4 +889,42 @@ impl FlatGraphBuilder {
             }
         }
     }
+
+    /// Operators with singleton references must be preceeded by a `persist()` in their pipeline,
+    /// to enable replay when the singleton updates.
+    fn ensure_singleton_referencers_succeed_persist(&mut self) {
+        // Resolve the singleton references for each node.
+        for node_id in self.flat_graph.node_ids().collect::<Vec<_>>() {
+            let GraphNode::Operator(operator) = self.flat_graph.node(node_id) else {
+                continue;
+            };
+            if operator.singletons_referenced.is_empty() {
+                continue;
+            }
+            let op_span = operator.span();
+            for (pred_edge, pred_node_id) in self
+                .flat_graph
+                .node_predecessors(node_id)
+                .collect::<Vec<_>>()
+            {
+                if let Some(op_inst) = self.flat_graph.node_op_inst(pred_node_id) {
+                    if "persist" == op_inst.op_constraints.name
```

Comment:
Are these "[stringly typed](https://en.wiktionary.org/wiki/stringly-typed)"? Should they be fixed (separately)?

Reply:
Good point, there is a better way

Reply:
I'd recommend creating an issue and dealing with it separately unless you think it's tiny and you can take care of it as soon as possible.

Reply:
Yeah turns out it is a bit more complicated that I thought